### PR TITLE
Fix grafana_datasource fails when name with spaces

### LINKED ISSUE #44154: Error when used Grafana datasource name with spaces
##### SUMMARY

Error from Garafan API when used datasource name with space(s).

##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME

grafana_datasource

##### ANSIBLE VERSION

```
ansible 2.6.1
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /bin/ansible
  python version = 2.7.5 (default, Aug  4 2017, 00:39:18) [GCC 4.8.5 20150623 (Red Hat 4.8.5-16)]
```

##### CONFIGURATION

DEFAULT_ASK_PASS(/etc/ansible/ansible.cfg) = True
DEFAULT_BECOME(/etc/ansible/ansible.cfg) = True
DEFAULT_BECOME_ASK_PASS(/etc/ansible/ansible.cfg) = True
DEFAULT_SSH_TRANSFER_METHOD(/etc/ansible/ansible.cfg) = piped
HOST_KEY_CHECKING(/etc/ansible/ansible.cfg) = False

##### OS / ENVIRONMENT

ansible host: CentOS 7 (3.10.0-693.5.2.el7.x86_64)
remote host: CentOS 7 (3.10.0-693.5.2.el7.x86_64)
Grafana version: 4.6.2-1 x86_64

##### STEPS TO REPRODUCE

Just use Grafana datasource name with space(s) in grafana_datasource module.

```yaml
---
- hosts: all

  tasks:
    - name: "Create Grafana datasource (PostgreSQL)"
      grafana_datasource:
        name: "name with spaces"
        grafana_url: "http://grafana-host:3000/"
        grafana_api_key: "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
        org_id: "1"
        ds_type: "postgres"
        url: "postgres-host:5432"
        database: "postgres"
        user: "postgres"
        password: "******"
        sslmode: "disable"
        state: "present"
      register: grafana_datasource_create
      failed_when:
        - "'Datasource added' not in grafana_datasource_create.msg"
        - "'already exists' not in grafana_datasource_create.msg"
```

##### EXPECTED RESULTS

Grafana datasource added.

##### ACTUAL RESULTS

HTTP Error 400: Bad Request

```
PLAY [all] ****************************************************************************************************************************************


TASK [Gathering Facts] ****************************************************************************************************************************

ok: [localhost]

TASK [Create Grafana datasource (PostgreSQL)] ****************************************************************************************************

fatal: [localhost]: FAILED! => {"changed": false, "msg": "error <class '__main__.GrafanaAPIException'> : Unable to get datasource name with spaces : {'body': '<!DOCTYPE HTML PUBLIC \"-//IETF//DTD HTML 2.0//EN\">\\n<html><head>\\n<title>400 Bad Request</title>\\n</head><body>\\n<h1>Bad Request</h1>\\n<p>Your browser sent a request that this server could not understand.<br />\\n</p>\\n</body></html>\\n', 'date': 'Tue, 14 Aug 2018 13:50:47 GMT', 'content-length': '226', 'status': 400, 'msg': 'HTTP Error 400: Bad Request', 'url': 'http://grafana-host:3000/api/datasources/name/name with spaces', 'content-type': 'text/html; charset=iso-8859-1', 'connection': 'close', 'server': 'Apache/2.4.6 (CentOS) PHP/5.4.16'} "}
        to retry, use: --limit @/opt/ansible/grafana-datasource.retry

PLAY RECAP ****************************************************************************************************************************************
```

### PR DESCRIPTION
**Pull request by martinwangjian**
##### SUMMARY

Fixes #44154

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

grafana_datasource
##### ANSIBLE VERSION

```paste below
ansible 2.7.1
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python3.6/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 3.6.6 (default, Jul 19 2018, 16:29:00) [GCC 7.3.1 20180303 (Red Hat 7.3.1-5)]
```

### CODE REVIEW

**Comment by gundalow** [comment]
@martinwangjian Thank you for your PR, your first I believe. Welcome to the Ansible Community!
Change looks sensible, though I don't have access to grafana to test.


@nikolaev-rd  does this fix the issue you reported in https://github.com/ansible/ansible/issues/44154

@seuf and @afsalrahim I believe you've worked on this module before, does this change look good?

File: lib/ansible/modules/monitoring/grafana_datasource.py [hunk 1]
```diff
@@ -310,7 +311,7 @@ def grafana_headers(module, data):
 def grafana_datasource_exists(module, grafana_url, name, headers):
     datasource_exists = False
     ds = {}
-    r, info = fetch_url(module, '%s/api/datasources/name/%s' % (grafana_url, name), headers=headers, method='GET')
+    r, info = fetch_url(module, '%s/api/datasources/name/%s' % (grafana_url, urlencode(name)), headers=headers, method='GET')
```
Comment (seuf) [inline lib/ansible/modules/monitoring/grafana_datasource.py]: urlencode take a dict as parameter.
you should replace this with `urlencode({"name": name})`

File: lib/ansible/modules/monitoring/grafana_datasource.py [hunk 2]
```diff
@@ -468,7 +469,7 @@ def grafana_delete_datasource(module, data):
     result = {}
     if datasource_exists is True:
         # delete
-        r, info = fetch_url(module, '%s/api/datasources/name/%s' % (data['grafana_url'], data['name']), headers=headers, method='DELETE')
+        r, info = fetch_url(module, '%s/api/datasources/name/%s' % (data['grafana_url'], urlencode(data['name'])), headers=headers, method='DELETE')
```
Comment (seuf) [inline lib/ansible/modules/monitoring/grafana_datasource.py]: same here

**Review by seuf — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/monitoring/grafana_datasource.py (hunk 2 above)
Comment (martinwangjian) [reply lib/ansible/modules/monitoring/grafana_datasource.py]: I use quote_plus now

File: lib/ansible/modules/monitoring/grafana_datasource.py (hunk 1 above)
Comment (martinwangjian) [reply lib/ansible/modules/monitoring/grafana_datasource.py]: I use quote_plus now

**Comment by seuf** [comment]
Hi. I stil lhave an issue with your fix. I can create a datasource with a space, but It's not idempotent.
When a re-run the playbook, i got an error :
```
fatal: [localhost]: FAILED! => {"changed": false, "msg": "error <class '__main__.GrafanaAPIException'> : Unable to create the new datasource influxdb datasource : 409 - {'url': 'https://grafana.company.com/api/datasources', 'content-length': '55', 'content-type': 'application/json', 'date': 'Thu, 22 Nov 2018 14:09:54 GMT', 'connection': 'close', 'msg': 'HTTP Error 409: Conflict', 'body': b'{\"message\":\"Data source with same name already exists\"}', 'status': 409}. "}
```
I think the `quote_plus` method does not format correctly the datasource name to test if it exists.

**Comment by gundalow** [comment]
@martinwangjian Hi, would you be able to take a look at the above comment from @seuf

**Comment by rrey** [comment]
Hi @martinwangjian,

A lot of your PRs were hanging !
I tested your patch and the `quote_plus` function is not formatting the url properly for grafana.

I tested with the `quote` function and grafana answers as expected.
Can you push the fix on a rebase on devel ?

**Comment by rrey** [comment]
@gundalow this issue can be closed in favor of #63987.

**Comment by rrey** [comment]
close_me

**Pull request closed without merging**